pymeasure/instruments/mksinst/mks937b.py

```diff
@@ -0,0 +1,291 @@
+                 "Convection Pirani": "CP",
+                 "Capacitance Manometer": "CM",
+                 "unknown (MB)": "MB",  # manual does not specify
+                 "unknown (GB)": "GB",  # manual does not specify
+                 "unknown (24)": "24",  # manual does not specify
+                 "no gauge": "NG",
+                 "no connection": "NC"}
+
+_ion_gauge_status = {"Wait": "W",
+                     "Off": "O",
+                     "Protect": "P",
+                     "Degas": "D",
+                     "Control": "C",
+                     "Rear panel Ctrl off": "R",
+                     "HC filament fault": "H",
+                     "No gauge": "N",
+                     "Good": "G",
+                     "NOT_IONGAUGE": "NAK152",
+                     }
+
+
+class MKS937B(Instrument):
+    """ MKS 937B vacuum gauge controller
+
+    :param resource_name: pyvisa resource name of the instrument
+    :param address: device address included in every message to the instrument
+                    (default=253)
+    :param kwargs: Any valid key-word argument for Instrument
+    """
+
+    def __init__(self, resource_name, address=253, **kwargs):
+        kwargs.setdefault("preprocess_reply", self._extract_reply)
+        kwargs.setdefault("write_termination", ";FF")
+        kwargs.setdefault("read_termination", ";")  # in reality its ";FF"
+        # which is, however, invalid for pyvisa. Therefore extra bytes have to
+        # be read in the read() method.
+        super().__init__(
+            resource_name,
+            "MKS 937B vacuum gauge controller",
+            includeSCPI=False,
+            **kwargs
+        )
+        self.address = address
+        # compiled regular expression for finding numerical values in reply strings
+        self._re_fullresponse = re.compile(fr"@{self.address:03d}(.*)")
+        self._re_response_value = re.compile(fr"@{self.address:03d}ACK(.*)")
+
+    def _extract_reply(self, reply):
+        """ preprocess_reply function which tries to extract <Response> from
+        '@<aaa>ACK<Response>;FF'. If <Response> can not be identified the orignal string
+        is returned.
+        :param reply: reply string
+        :returns: string with only the response, or the original string
+        """
+        rvalue = self._re_response_value.search(reply)
+        if rvalue:
+            return rvalue.groups()[0]
+        else:
+            full_reply = self._re_fullresponse.search(reply)
+            if full_reply:
+                return full_reply.groups()[0]
+        return reply
+
+    def _command_process(self, cmd):
+        """
+        create command string by adding device address
+        """
+        return f"@{self.address:03d}{cmd}"
+
+    def _check_extra_termination(self):
+        """
+        Check the read termination to correspond to the protocol
+        """
+        t = self.adapter.read_bytes(2)  # read extra termination chars 'FF'
```

Comment:
@bmoneke that approach might be more challenging to test with our current testing setup (either ProtocolAdapter or the Mock when checking all instruments). I guess we will see on the next CI run :D

Reply:
You have to make the communication pairs longer, but it should work.
E.g.
comm_pairs = [(b"writing something", b"the response"), (None, b"FF")]

Remember, that you must not use the termination_characters defined at adapter level (i.e. ";FF" for writing and ";" for reading).

Reply:
I tried to understand how the protocol test could work for this device but did not succeed. From the [documentation](https://github.com/pymeasure/pymeasure/blob/master/docs/dev/adding_instruments.rst) I assumed I can write something like:

```
import pytest

from pymeasure.test import expected_protocol
from pymeasure.instruments.mksinst.mks937b import MKS937B

def test_pressure():
    """Verify the communication of the voltage getter."""
    with expected_protocol(
        MKS937B,
        [("@253PR1?", "@253ACK1.10e-9"),
         (None, b"FF")],
    ) as inst:
        assert inst.pressure1 == 1.1e-9
```

but this failed with:
` AssertionError: assert '@253ACK1.10e-9' == 1.1e-09`

Can you give me a hint on how this is done for this instrument? I guess it makes sense I write few basic tests for this to be included in this PR

Reply:
The test looks right and I guess it does its job, exposing coding errors 😀: `inst.pressure1` returns the string the device responded and probably you made somewhere in your code a mistake (or values does not use the `preprocess_reply`).
Maybe your `_extract_reply` does not work as expected. I'd write a test for that method as well.

Reply:
Not sure i can agree here. testing the same with the hardware seems to work
```
>>> import pymeasure.instruments
>>> inst = pymeasure.instruments.mksinst.MKS937B("ASRL3::INSTR")
>>> inst.pressure1
1.6e-08
>>>
```
So with the hardware all functions seem to be called as I intend it. Something seems different in the test case?

Reply:
As extra info here the raw hardware communication for the same command:
```
>>> import pyvisa
>>> rm = pyvisa.ResourceManager()
>>> inst = rm.open_resource("ASRL3::INSTR")
>>> inst.write("@253PR1?;FF")
13
>>> inst.read_bytes(18)
b'@253ACK1.50E-08;FF'
>>>
```

Reply:
I found the problem: The ProtocolAdapter does not get the global preprocess_reply configuration, as only a newly created VISAAdapter gets that information. (I'm reworking the adapters, so that's going to change in the future).

As a workaround you can add `inst.adapter.preprocess_reply = inst._extract_reply` directly after the colon in the test. That way you add the preprocess to the adapter and the test succeeds.
In theory you can hand parameters to the ProtocolAdapter if you define them after the communication pairs. In you case, that does not work, as the preprocess_reply is a method of the instrument instance.

Reply:
thanks with the workaround the tests succeed. I have added some to test the different types of properties of this new device.

Reply:
Thanks for the tests. It makes future maintenance easier. You have some style errors (one instead of two newlines etc.), see the file with the comments made by CI.

Reply:
Great troubleshooting, folks! Thanks for supplying complete information; thanks for helping out getting a good test running! "This is the way!" :D